### PR TITLE
WIP: leverage Base reduction functions

### PR DESCRIPTION
**Pull request by simonbyrne**
This is something I have been meaning to do for a while. Basically the idea is to define a kbn addition operator, and define all the reductions in terms of that.

### CODE REVIEW

File: src/KahanSummation.jl [hunk 1]
```diff
@@ -29,6 +29,59 @@ else
     promote_sys_size_add(x::T) where {T} = Base.r_promote(+, zero(T)::T)
 end
 
+
+import Base.TwicePrecision
+
+
+function plus_kbn(x::T, y::T) where {T}
```
Comment (ararslan) [inline src/KahanSummation.jl]: Should these be restricted to `T<:Real`? Also do they necessarily have to be the same type? Using `+` as we do here should take care of any promotion that needs to happen.

File: src/KahanSummation.jl [hunk 2]
```diff
@@ -29,6 +29,59 @@ else
...
+"""
+sum_kbn(f, X) = singleprec(mapreduce(f, plus_kbn, X))
+sum_kbn(X) = sum_kbn(identity, X)
+
+
+
+
+
+
+
```
Comment (ararslan) [inline src/KahanSummation.jl:103]: This is a lot of blank lines. Could you reduce it to just one, maybe two?

**Comment by ararslan** [comment]
This looks really cool! Is there any noticeable change in performance?

File: src/KahanSummation.jl (hunk 2 above)
Comment (simonbyrne) [reply src/KahanSummation.jl:103]: No. I need those for caching :trollface:

**Comment by simonbyrne** [comment]
AHHHHH!!!!!

This doesn't work on 0.7 because `r_promote` was removed (https://github.com/JuliaLang/julia/pull/22825). The recommendation seems to be that you combine the promotion into the mapping function, but that will be pretty inefficient in this case, as it will end up calling `plus_kbn(::TwicePrecisionN, ::TwicePrecisionN)` instead of `plus_kbn(::TwicePrecisionN, ::Float64)`.

**Comment by mgr327** [comment]
Would it be possible to re-activate/implement that pull request?

**Comment by JeffreySarnoff** [comment]
We do not care about support for Julia 0.7.  If you can get this work as intended for Julia "1.6+", we are open to including it.

**Comment by asinghvi17** [comment]
I just looked at this code again and it seems to be fine in v1.11.

```
# this implementation
julia> @be sum_kbn($(rand(1000)))
Benchmark: 3348 samples with 16 evaluations
 min    1.677 μs
 median 1.734 μs
 mean   1.738 μs
 max    7.698 μs

# the implementation on master
julia> @be KahanSummation.sum_kbn($(rand(1000)))
Benchmark: 3336 samples with 16 evaluations
 min    1.682 μs
 median 1.698 μs
 mean   1.740 μs
 max    4.336 μs

# compared to current Base pairwise summation, slower of course
julia> @be sum($(rand(1000)))
Benchmark: 3198 samples with 290 evaluations
 min    98.131 ns
 median 98.707 ns
 mean   100.700 ns
 max    268.390 ns
```
I attached a timer output and it never calls the method for two TwicePrecisions.  (Annotations mine)
```
julia> to
────────────────────────────────────────────────────────────────────
                           Time                    Allocations      
                  ───────────────────────   ────────────────────────
Tot / % measured:      2.13s /   0.8%           3.68MiB /   0.0%    

Section                           ncalls     time    %tot     avg     alloc    %tot      avg
───────────────────────────────────────────────────────────────────────────────
+(::Float64, ::TwicePrecisionN)     1.72M   17.9ms   99.9%  10.4ns     0.00B     - %    0.00B
+(::Float64, ::Float64)             1.72k   17.4μs    0.1%  10.1ns     0.00B     - %    0.00B
───────────────────────────────────────────────────────────────────────────────
```

But this seems mergeable as is.

cc @simonbyrne @JeffreySarnoff

**Pull request closed without merging**